sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/cryptography/EncryptedBlockBlobAsyncClient.java

```diff
@@ -0,0 +1,403 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.blob.cryptography;
+
+import com.azure.core.http.HttpPipeline;
+import com.azure.core.http.rest.Response;
+import com.azure.core.http.rest.SimpleResponse;
+import com.azure.core.http.rest.VoidResponse;
+import com.azure.core.implementation.util.FluxUtil;
+import com.azure.core.util.Context;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.storage.blob.BlobAsyncClient;
+import com.azure.storage.blob.BlockBlobAsyncClient;
+import com.azure.storage.blob.UploadBufferPool;
+import com.azure.storage.blob.implementation.AzureBlobStorageImpl;
+import com.azure.storage.blob.models.AccessTier;
+import com.azure.storage.blob.models.AccessTierOptional;
+import com.azure.storage.blob.models.BlobAccessConditions;
+import com.azure.storage.blob.models.BlobHTTPHeaders;
+import com.azure.storage.blob.models.BlockBlobItem;
+import com.azure.storage.blob.models.BlockLookupList;
+import com.azure.storage.blob.models.CpkInfo;
+import com.azure.storage.blob.models.LeaseAccessConditions;
+import com.azure.storage.blob.models.Metadata;
+import com.azure.storage.common.Constants;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.AsynchronousFileChannel;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.Base64;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import static com.azure.core.implementation.util.FluxUtil.withContext;
+import static com.azure.storage.blob.PostProcessor.postProcessResponse;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class EncryptedBlockBlobAsyncClient extends BlobAsyncClient {
+
+    static final int BLOB_DEFAULT_UPLOAD_BLOCK_SIZE = 4 * Constants.MB;
+    private static final int BLOB_MAX_UPLOAD_BLOCK_SIZE = 100 * Constants.MB;
+    private final ClientLogger logger = new ClientLogger(EncryptedBlockBlobAsyncClient.class);
+
+    private final BlobEncryption encryptionPolicy;
+
+    /**
+     * Package-private constructor for use by {@link EncryptedBlobClientBuilder}.
+     */
+    EncryptedBlockBlobAsyncClient(AzureBlobStorageImpl constructImpl, String snapshot, CpkInfo cpk,
```

Comment:
Opinions on having encryptedblockblobasyncclient instead just wrap a blockblobasyncclient (that way the commitblock and stageblock will not be repeated in this class)

Reply:
Wrapping BlockBlobAsyncClient should work if the method internals are the same outside of this client having an additional policy.

Reply:
It's not just an additional policy, it also encrypts the data before sending it on it's way through the pipeline